### DISCUSSION TITLE
Feature Request: C-API:  Add support for constructing a ListVector

### DISCUSSION THREAD

**Discussion by eddyxu** [duckdb/duckdb#6152] [Ideas]
Why do you want this feature?

We are building a duckdb extension to scan [Lance Format](https://github.com/eto-ai/lance) via a TableFunction, in Rust, which calls duckdb C-API.

To load nested fields into duckdb, it is needed to construct a ListVector, and append values to it. 

However, because `list_entry_t` struct is not exposed in C API, it seems to be difficult to build 
a `ListVector` via the following two C APIs.

```c
DUCKDB_API duckdb_vector duckdb_list_vector_get_child(duckdb_vector vector);
DUCKDB_API void *duckdb_vector_get_data(duckdb_vector vector);
```

Desired interface

```c
typedef list_entry_t {
...
} * list_entry_t;
DUCKDB_API void duckdb_list_vector_append(duckdb_vector list_vector, duckdb_vector child, /* offset, length */);
```

**Comment by Tishj** [comment]
I'm not entirely sure how you envision the `duckdb_list_vector_append` to work
Essentially the ListVector is nothing more than a wrapper on another "child" Vector, with 1 set of `list_entry_t` per row.
The `offset` of the entry indicates the index in the `child` vector.
The `length` of the entry indicates how many sequential indices from `offset` are part of this row.

There is only ever 1 child.

If you want a more friendly, but less efficient API we could look into adding support for lists to the Value API.

**Reply by eddyxu** [reply]
Does `ListVector` in C++ do a Copy from the `source` vector ( source[offset: offset + length]) to the ListVector's auxiliary child?

I was thinking about workflow like:

```
# pesudo-code
# for a list of  [[1,2,3], [4,5,6], [7,8,9]]

ArrowInt32Array child = [1, 2, 3, 4, 5, 6, 7, 8, 9];  // Get this Arrow array
duckdb_vector child_vector = ...;
duckdb_vector list = ...;
duckdb_list_vector_append(list, child_vector, 0, 3); // Which internally call ListVector::Append or ListVector::Push
duckdb_list_vector_append(list, child_vector, 3, 3);
duckdb_list_vector_append(list, child_vector, 6, 3);
```

Would an efficient way to implement it looks like the following?

```
duckdb_vector list = ...
duckdb_vector child_vector = duckdb_list_vector_get_child(list);
memcpy(duckdb_vector_get_data(child_vector), child.data(), 9 * sizeof(int));
duckdb_list_vector_set(list, 0 /* offset */, 3 /* length */);
duckdb_list_vector_set(list, 3 /* offset */, 3 /* length */);
duckdb_list_vector_set(list, 6 /* offset */, 3 /* length */);
```

**Reply by Mytherin** [reply]
Yes, the second method looks a lot more efficient, and is more similar to how this would be implemented in the C++ API. We could enable this by adding the following methods, plus adding the `duckdb_list_entry` struct.

```cpp
DUCKDB_API void duckdb_list_vector_reserve(duckdb_vector vector, idx_t capacity);
DUCKDB_API void duckdb_list_vector_set_size(duckdb_vector vector, idx_t size);

typedef struct {
    uint64_t offset;
    uint64_t length;
} duckdb_list_entry;
```

Then you could construct a list vector like so:

```cpp
duckdb_vector list = ...

// reserve space for 9 elements in the the child vector
duckdb_list_vector_reserve(list, 9);
duckdb_list_entry *entries = (duckdb_list_entry *) duckdb_vector_get_data(list);
// set the offsets of the list entries
entries[0].offset = 0;
entries[0].length = 3;
entries[1].offset = 3;
entries[1].length = 3;
entries[2].offset = 3;
entries[2].length = 3;

// set the child data
duckdb_vector child = duckdb_list_vector_get_child(list);
int32_t *child_data = (int32_t *) duckdb_vector_get_data(child);
child_data[0] = 1;
child_data[1] = 2;
child_data[2] = 3;
child_data[3] = 4;
child_data[4] = 5;
child_data[5] = 6;
child_data[6] = 7;
child_data[7] = 8;
child_data[8] = 9;

// sets the number of entries in the child of the list
duckdb_list_vector_set_size(9);
```

**Comment by eddyxu** [comment]
KK thanks @Tishj and @Mytherin 

Worked on a quick PR to implement the discussed solution above

https://github.com/duckdb/duckdb/pull/6155

Would appreciate to have some reviews!
